File: src/main/java/com/starshootercity/abilities/VisibleAbility.java
package com.starshootercity.abilities;

import com.starshootercity.OriginSwapper;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public interface VisibleAbility extends Ability {
    @NotNull List<OriginSwapper.LineData.LineComponent> getDescription();
    @NotNull List<OriginSwapper.LineData.LineComponent> getTitle();
}

File: core/src/main/kotlin/ru/turbovadim/abilities/VisibleAbility.kt
package ru.turbovadim.abilities

import ru.turbovadim.AddonLoader.getTextFor
import ru.turbovadim.OriginSwapper.LineData.Companion.makeLineFor
import ru.turbovadim.OriginSwapper.LineData.LineComponent

interface VisibleAbility : Ability {
    val description: List<LineComponent>
    val title: List<LineComponent>

    val usedDescription: List<LineComponent>
        get() {
            val keyText = getKey().toString().replace(":", ".")
            val text = getTextFor("power.$keyText.description")
            return if (text != null) {
                makeLineFor(text, LineComponent.LineType.DESCRIPTION)
            } else {
                description
            }
        }

    val usedTitle: List<LineComponent>
        get() {
            val keyText = getKey().toString().replace(":", ".")
            val text = getTextFor("power.$keyText.name")
            return if (text != null) {
                makeLineFor(text, LineComponent.LineType.TITLE)
            } else {
                title
            }
        }
}
